src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,419 @@
+
+ Outputs: Returns the address of the allocated symbol
+
+ Purpose: Create a symbol for a pointer to point to
+
+\*******************************************************************/
+
+exprt symbol_factoryt::allocate_object(
+  code_blockt &code,
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const bool static_lifetime)
+{
+  // Add the following code to init_code:
+  //   <target_expr> = &tmp$<temporary_counter>
+  // Store the following code in the input recorder:
+  //   INPUT("tmp$<temporary_counter>", tmp$<temporary_counter>);
+
+  symbolt &aux_symbol=c_new_tmp_symbol(symbol_table, loc, allocate_type);
+  aux_symbol.is_static_lifetime=static_lifetime;
+
+  exprt object=aux_symbol.symbol_expr();
+  exprt aoe=address_of_exprt(object);
+
+  const typet &allocate_type_resolved=ns.follow(allocate_type);
+  const typet &target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+  if(cast_needed)
+  {
+    aoe=typecast_exprt(aoe, target_expr.type());
+  }
+
+  code_assignt assign(target_expr, aoe);
+  assign.add_source_location()=loc;
+  code.add(assign);
+
+  input_recorder.record_input(aux_symbol.base_name, object);
+
+  return aoe;
+}
+
+/*******************************************************************\
+
+Function: symbol_factoryt::gen_nondet_init
+
+ Inputs:
+  init_code - The code block to add code to
+  expr - The expression which we are generating a non-determinate value for
+
+ Outputs:
+
+ Purpose: Creates a nondet for expr, including calling itself recursively to
+          make appropriate symbols to point to if expr is a pointer.
+
+\*******************************************************************/
+
+void symbol_factoryt::gen_nondet_init(
+  code_blockt& init_code,
+  const exprt &expr)
+{
+  const typet &type=ns.follow(expr.type());
+
+  if(type.id()==ID_pointer)
+  {
+    // dereferenced type
+    const pointer_typet &pointer_type=to_pointer_type(type);
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_blockt non_null_inst;
+
+    // Note, the allocated object needs to be static. E.g., with a pointer to a
+    // pointer to a pointer temporary variables are created in else clauses,
+    // and thus go out of scope before the function is called.
+    exprt allocated=
+      allocate_object(non_null_inst, expr, subtype, true);
+
+    exprt init_expr;
+    if(allocated.id()==ID_address_of)
+    {
+      init_expr=allocated.op0();
+    }
+    else
+    {
+      init_expr=dereference_exprt(allocated, allocated.type().subtype());
+    }
+    gen_nondet_init(non_null_inst, init_expr);
+
+    if(assume_non_null)
+    {
+      // Add the following code to init_code:
+      // <expr> = <aoe>;
+      concatenate_code_blocks(init_code, non_null_inst);
+    }
+    else
+    {
+      // Add the following code to init_code:
+      //           IF !(NONDET(_Bool) == FALSE) THEN GOTO <label1>
```

Comment:
I'm not sure exactly what you mean. Do you mean that changing the condition of the if-then-else clause means that this line changes? I ended up having to back out that change because it wouldn't run with just --cover branch, without --show-goto-functions.